src/types/list.md

```diff
@@ -1 +1,66 @@
+
+The other essential component to lists is what is known as the `::` operator, referred to as "cons".
+```sml
+(op ::) : int * int list -> int list
+```
+
+Cons takes in an int and an int list, and prepends the int to the front of the given int list. For instance, we have that `1 :: [2, 3]` steps to `[1, 2, 3]`. Additionally, cons is _right-associative_. This means that in a continuous stream of applications of cons, they are evaluated from _right-to-left_. This means that `1 :: 2 :: 3 :: []` is implicitly denoting `1 :: (2 :: (3 :: []))`, as the calls to cons associate to the right. So, similarly to before, a `1 :: [2, 3]`, `[1, 2, 3]`, and `1 :: 2 :: 3 :: []` are exactly equivalent, and denote the same list.
+
+Cons is also a _constructor_, meaning that it can be used to pattern match and deconstruct lists. As such, we can write basic functions to compute the length of a list as follows:
+```sml
+(* length : int list -> int *)
+(* REQUIRES: true *)
+(* ENSURES: length L ==> the length of L *)
+fun length ([] : int list) : int = 0
+  | length (x::xs : int list) : int = 1 + length xs
+```
+Given a non-empty list, this function simply binds the first element of the list to the identifier `x`, discards it, and then recursively calls `length` to find the length of the remaining list `xs`, adding 1 to the result.
+
+The definition of an int list thus corresponds to:
+```sml
+datatype int list = [] | :: of int * int list
+```
+where an int list can either be the empty list `[]`, or it can be `::` of a first element and the rest of the list (where `::` is an infix operator, so instead of being written as `::(x, xs)`, we have `x :: xs`).
+
+## Motivation
+
+Compared to other data structures, lists seem to have numerous disadvantages. As mentioned previously, they do not possess constant-time indexing like arrays in other languages - there is no way to instantly get the ith element of a list easily. Instead, you must "cons off" all the elements in front of that item in order to retrieve it - if you want to remove that item from the list, then you have to put the preceding elements _back_ as well (and in the right order!). Lists are also inherently sequential - you cannot access multiple elements at one time. 
+
+The reason why we choose lists is that they have very nice mathematical properties. These "disadvantages" in the previous paragraph become strengths, when viewed in a certain manner. Lists are powerful for their simple, inductive definition (as shown in the previous section), which is sufficiently powerful to characterize many important principles in this class. Our hope is that, throughout this course, you can begin to see that there is an elegance in simplicity.
```

Comment:
These two paragraphs seem unconvincing to me. I'm not sure of a good way to immediately resolve that, but offhand I'd either avoid bringing it up at all, or mentioning some concrete ways that it isn't a disadvantage (persistence, desgining code flows such that it matters less, etc).